### ISSUE TITLE
What is adding this wrapper around my code?

### ISSUE THREAD

**Issue by RoxKilly** [systemjs/systemjs#1371]
**### SUMMARY  ###**

I'm writing an Angular 2 app in `TypeScript`, and using `SystemJS` and `Gulp` to deploy it. I'm facing a curious problem when trying to use sourcemaps. If I include the sourcemap inline, everything works. If I use an external map file, `SystemJS` seems to add a wrapper around my code that I don't understand, breaking the browser's ability to find the sourcemap. I would like to figure out what's actually happening and how I can fix it.

**### INLINE MAPS WORK ###**

**gulpfile.js** Here is the task that compiles TypeScript to JavaScript

    gulp.task('build-ts', function () {
        return gulp.src(appDev + '**/*.ts')
            .pipe(sourcemaps.init()) //gulp-sourcemaps
            .pipe(typescript(tsProject)) //gulp-typescript
            .pipe(sourcemaps.write()) //map will be inline
            .pipe(gulp.dest(appProd));
    });
The resulting JS file (**login.component.js**) on disk and in the browser ends like this:

    165| exports.LoginComponent = LoginComponent;
    166| 
    167| //# sourceMappingURL=data:application/json;base64,eyJ2Z...

The browser is able to use the inline map to let me see the source TypeScript file. No problem.

**### EXTERNAL MAPS ARE BROKEN ###**

**gulpfile.js** to generate external maps

    gulp.task('build-ts', function () {
        return gulp.src(appDev + '**/*.ts')
            .pipe(sourcemaps.init())
            .pipe(typescript(tsProject))
            .pipe(sourcemaps.write('.')) //map will be external
            .pipe(gulp.dest(appProd));
    });
The resulting JS file (**login.component.js**) *on disk* is the same as above except line 167:

    165| exports.LoginComponent = LoginComponent;
    166| 
    167| //# sourceMappingURL=login.component.js.map

The same file, when I view the source in the browser ends like this:

    165| exports.LoginComponent = LoginComponent;
    166| 
    167| //# sourceMappingURL=login.component.js.map
    168| 
    169| }).apply(__cjsWrapper.exports, __cjsWrapper.args);
    170| })(System, System);
    171| //# sourceURL=http://example.com/path/to/login.component.js

The map file *login.component.js.map* is properly generated in the same directory but Firefox never fetches it.  This wrapping breaks Firefox's ability to load the sourcemap. Why would the file generated by the gulp task be different from the same file once the browser loads it?

Am I asking for help in the wrong place?

Note: this is cross-posted on [StackOverflow](http://stackoverflow.com/questions/38495497/what-is-adding-this-wrapper-around-my-code)

**Comment by asapach** [comment]
I would recommend using `System.register()` format instead of CommonJS - that way it can be loaded without any wrappers. And set `scriptLoad: true` in your [`meta`](https://github.com/systemjs/systemjs/blob/master/docs/config-api.md#meta) to actually load it as a script tag - that way Firefox (and any other browser that understands sourcemaps) will play nice.

**Comment by RoxKilly** [comment]
@asapach thanks for the reply. I'm really sorry but all of that went over my head. I've never worked with SystemJS or CommonJS before so I'm not wise to the lingo (I was just given these files as part of a web tutorial I'm following). I will look into what `System.register()` is. This may not be relevant but just in case, I'll share:

I have a `systemjs.config.js` file in which I build `map` and `packages` array then do:
```
var config = {
    map: map,
    packages: packages
};
System.config(config);
```
In my `index.html`, I point to that config file with a `<script>` before I call ` System.import('app')`

Does any of that change what you wrote?

**Comment by asapach** [comment]
In your [`tsconfig.json`](https://www.typescriptlang.org/docs/handbook/tsconfig-json.html) set `"module": "system"` instead of default `"commonjs"`.

In your `packages` config add this:
```js
{
...
  app: {
    meta: {
      format: "register",
      scriptLoad: true
    }
  }
...
}
```

**Comment by RoxKilly** [comment]
I changed my `SystemJS` config object to
```
    var config = {
        map: map,
        packages: packages,
        meta: {
            format: "register",
            scriptLoad: true
        }
    };
```
By itself, that made no difference I noticed. I then also changed `"module": "commonjs"` to `"module": "system"` in tsconfig, but that broke the app altogether. All files downloaded fine (no HTTP error), but I received a generic `Error loading main.js` in the console.  This is an Angular 2 app setup according to [the quickstart tutorial](https://angular.io/docs/ts/latest/quickstart.html); I think the tutorial sets things up to use commonjs, so making the change to system isn't trivial for me.

I'm really curious though: as I pointed out in the OP, the generated JS file at the end of all gulp tasks looks differently from the file loaded in the browser. Why would this be? when is this wrapping added. It really puzzled me.

**Comment by asapach** [comment]
CommonJS cannot be natively executed in the browser, because it doesn't have the built-in globals like `require`, `process`, etc., so it needs to be wrapped in order to inject them.

Your config is incorrect: `meta` should be inside a package named `app`. See docs: https://github.com/systemjs/systemjs/blob/master/docs/config-api.md#packages

**Comment by RoxKilly** [comment]
In `systemjs.config.js` I changed the `config.packages` setting for my app from
```
'app':{ 
	main: 'main.js',
    defaultExtension: 'js'
}
```
to:
```
'app':{ 
	main: 'main.js',
    defaultExtension: 'js',
    meta: {
    	format: "register",
        scriptLoad: true
        }
}
```
I could discern no difference. I then also changed the module setting in `tsconfig.json` from `commonjs` to `system` and it broke the app just as in my last post.

@asapach you mentioned that commonjs needs to be wrapped because it cannot be natively executed. What's really puzzling though is that I see no wrapping when I use inline maps. I mean the only thing I do differently to run into this problem is change `sourcemaps.write()` to `sourcemaps.write('.')` which creates an external map. That's it. Why isn't there wrapping in the 1st case given that I'm using commonjs either way?  

Even more puzzling: why doesn't the wrapping appear once all the JS files are built?? I mean, if wrapping is a consequence of a `tsconfig` setting, I should see it once the `.ts` files are transpiled to `.js` right? As I said in the OP, on disk there is no wrapping. It's only when I look at the source code in the browser that I see wrapping.  When is it added?

**Comment by RoxKilly** [comment]
**Important update:**

It seems that the wrapping has nothing to do with why Firefox can't load the external maps. Rather, Firefox is unable to resolve relative paths in `sourceMappingURL`.  In my last excerpt above, if I change line 167 from:

    //# sourceMappingURL=login.component.js.map

to:

    //# sourceMappingURL=http://example.com/path/to/login.component.js.map

Then Firefox will load the map! Now I just need to figure out how to get `gulp-sourcemaps` to write the full URL when linking to external maps

**Comment by asapach** [comment]
> why doesn't the wrapping appear once all the JS files are built?

That's what `"module": "system"` does: it wraps the module into a function that can be executed.

**Comment by RoxKilly** [comment]
@asapach Ok. I've been doing a bad job of explaining what puzzles me. I'll try again. Please try to explain what you're trying to get me to understand just one more time.  Below is my understanding of the steps that turn my code into a Javascript app.

1.  I setup `tsconfig.json`, `systemjs.config.js`, and `gulpfile.js` on the dev machine
2.  I write my app in TypeScript
3.  I run a gulp task through plugins `gulp-sourcemaps` and `gulp-typescript`. This build task takes my source files and turns them into `.js` and `.js.map` files
4.  I push the JavaScript and map files to my server
5.  Browser makes an http request and downloads the `.js` and `.js.map` files
6.  Browser executes the JavaScript.
7.  I look at the JS source in browser dev tools

 What I've been describing as puzzling, surprising etc.. is that on step 4, I see no wrapping, so the files I push to the server don't have the wrapping.  But on step 7, I see the wrapping. At what step in this chain is the wrapping added? What adds it?

**Comment by asapach** [comment]
> What adds it?

SystemJS does - that's its job: converting other module formats (CommonJS, AMD, ES6 modules and plain global scripts) to something that can be executed in the browser according to the EcmaScript module spec. If you don't do the convertion work at build time (by transpiling to "system" format) it will happen at runtime - that's why you see the discrepancy.

**Comment by RoxKilly** [comment]
Interesting. Thanks a ton for your time.  

For reference to anyone who has the same problem.  Once I realized that my issue had nothing to do with the wrapping, and that Firefox simply doesn't resolve relative links in `SourceMappingURL`, the problem became simpler. I looked at the `gulp-sourcemaps` [documentation](https://www.npmjs.com/package/gulp-sourcemaps) to find out how to use full URL links.  I fixed my issue by making use of the `sourceMappingURL` parameter.  In `gulpfile.js`, I changed my typescript build task from:
